execution/tvf/watermark_generator.go

```diff
@@ -0,0 +1,123 @@
+package tvf
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/docs"
+	"github.com/cube2222/octosql/execution"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+type WatermarkGenerator struct {
+	source    execution.Node
+	timeField octosql.VariableName
+	offset    execution.Expression
+}
+
+func NewWatermarkGenerator(source execution.Node, timeField octosql.VariableName, offset execution.Expression) *WatermarkGenerator {
+	return &WatermarkGenerator{
+		source:    source,
+		timeField: timeField,
+		offset:    offset,
+	}
+}
+
+func (w *WatermarkGenerator) Document() docs.Documentation {
+	panic("implement me")
+}
+
+func (w *WatermarkGenerator) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, *execution.ExecOutput, error) {
+	tx := storage.GetStateTransactionFromContext(ctx)
+	sourceStreamID, err := execution.GetSourceStreamID(tx.WithPrefix(streamID.AsPrefix()), octosql.MakePhantom())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source stream ID")
+	}
+
+	source, _, err := w.source.Get(ctx, variables, sourceStreamID) // we don't need execOutput here since we become new watermark source
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source")
+	}
+
+	offset, err := w.offset.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark offset")
+	}
+	if offset.GetType() != octosql.TypeDuration {
+		return nil, nil, errors.Errorf("invalid watermark offset: %v", offset)
```

Comment:
invalid watermark offset _type_. And after the value, please add the result of offset.GetType() too.